slsReceiverSoftware/src/slsReceiverTCPIPInterface.cpp

```diff
@@ -682,1580 +547,793 @@ int slsReceiverTCPIPInterface::set_roi(sls::ServerInterface2 &socket) {
-						FILE_LOG(logERROR) << mess;
-					} else {
-						strcpy(retvals[0],temp.c_str());
-						FILE_LOG(logINFO) << "Receiver MAC Address: " << retvals[0];
-					}
-				}
 			}
+			impl()->setEthernetInterface2(eth);
 
-			// 2nd interface
-			if (myDetectorType == JUNGFRAU && numInterfaces == 2) {
-				receiver->setUDPPortNumber2(port2);
-				FILE_LOG(logINFO) << "Receiver UDP IP 2: " << ip2;
-				// get eth
-				temp = sls::IpToInterfaceName(ip2);
-				if (temp == "none"){
+			//get mac address
+			if (ret != FAIL) {
+				temp = sls::InterfaceNameToMac(eth).str();
+				if (temp=="00:00:00:00:00:00") {
 					ret = FAIL;
-					strcpy(mess, "Failed to get 2nd ethernet interface or IP \n");
+					strcpy(mess,"failed to get 2nd mac adddress to listen to\n");
 					FILE_LOG(logERROR) << mess;
 				} else {
-					char eth[MAX_STR_LENGTH] = {""};
-					memset(eth, 0, MAX_STR_LENGTH);
-					strcpy(eth, temp.c_str());
-					// if there is a dot in eth name
-					if (strchr(eth, '.') != nullptr) {
-						strcpy(eth, "");
-						ret = FAIL;
-						sprintf(mess, "Failed to get 2nd ethernet interface from IP. Got %s\n", temp.c_str());
-						FILE_LOG(logERROR) << mess;
-					}
-					receiver->setEthernetInterface2(eth);
-
-					//get mac address
-					if (ret != FAIL) {
-						temp = sls::InterfaceNameToMac(eth).str();
-						if (temp=="00:00:00:00:00:00") {
-							ret = FAIL;
-							strcpy(mess,"failed to get 2nd mac adddress to listen to\n");
-							FILE_LOG(logERROR) << mess;
-						} else {
-							strcpy(retvals[1],temp.c_str());
-							FILE_LOG(logINFO) << "Receiver MAC Address 2: " << retvals[1];
-						}
-					}
+					strcpy(retvals[1],temp.c_str());
+					FILE_LOG(logINFO) << "Receiver MAC Address 2: " << retvals[1];
 				}
 			}
-
-			// set the number of udp interfaces (changes number of threads and many others)
-			if (myDetectorType == JUNGFRAU && receiver->setNumberofUDPInterfaces(numInterfaces) == FAIL) {
-				ret = FAIL;
-				sprintf(mess, "Failed to set number of interfaces\n");
-				FILE_LOG(logERROR) << mess;
-			}
 		}
 	}
-	return socket.sendResult(true, ret, retvals, sizeof(retvals), mess);
-}
-
-
 
+	// set the number of udp interfaces (changes number of threads and many others)
+	if (myDetectorType == JUNGFRAU && impl()->setNumberofUDPInterfaces(numInterfaces) == FAIL) {
+		throw RuntimeError("Failed to set number of interfaces");
+	}
+	return socket.sendResult(ret, retvals, sizeof(retvals), mess);
+}
 
-int slsReceiverTCPIPInterface::set_timer(sls::ServerInterface2 &socket) {
-	ret = OK;
+int slsReceiverTCPIPInterface::set_timer(Interface &socket) {
 	memset(mess, 0, sizeof(mess));
 	int64_t index[2] = {-1, -1};
 	int64_t retval = -1;
-
-	// get args, return if socket crashed, ret is fail if receiver is not null
-	socket.receiveArg(ret, mess, &index, sizeof(index));
-	if(receiver == nullptr){
-		NullObjectError(ret, mess);
-		return FAIL;
-	}
-
-	// base object not null
-	if (ret == OK) {
+	socket.receiveArg(index);
```

Comment:
Could you use socket.receive? or is it not possible since it is an array?

Reply:
I would need to rewrite it to be an std::array<int64_t,2> instead of int64_t index[2]. 
The one that we have now decays to a pointer